File: lib/utils/table_utils.dart
import 'package:total_pos/models/table.dart';

List<List<T?>> rotate2dMatrix<T>(List<List<T?>> matrix) {
  final List<List<T?>> newList2d = List.generate(
      matrix[0].length, (_) => List.generate(matrix.length, (_) => null));
  for (int i = 0; i < matrix.length; i++) {
    for (int j = 0; j < matrix[i].length; j++) {
      newList2d[j][matrix.length - 1 - i] = matrix[i][j];
    }
  }
  return newList2d;
}

List<Table?> tablesDesktoTablesMobile(List<Table?> matrix) {
  final List<List<Table?>> list2d =
      List.generate(5, (_) => List.generate(10, (_) => null));
  for (final table in matrix) {
    if (table == null) continue;
    list2d[table.posX - 1][table.posY - 1] = table;
  }
  final rotateList = rotate2dMatrix(list2d);
  final linearList = List<Table?>.filled(50, null);
  for (var i = 0; i < rotateList.length; i++) {
    for (var j = 0; j < rotateList[i].length; j++) {
      linearList[(i * 5) + (j)] = rotateList[i][j];
    }
  }
  return linearList;
}

File: src/utils/tableUtils.js
export const orderTables = (tables, isVertical = false) => {
  const tables2d = Array.from({ length: 5 }, () => Array.from({ length: 10 }, () => undefined))
  for (const table of tables) {
    tables2d[table.pos_y - 1][table.pos_x - 1] = table
  }
  const t2d = isVertical ? rotateMatrix(tables2d) : tables2d
  const tablesOrdered = []
  for (const row of t2d) {
    for (const t of row) {
      tablesOrdered.push(t)
    }
  }
  return tablesOrdered
}

export const rotateMatrix = (matrix) => {
  const newList2d = Array.from({ length: matrix[0].length }, () => Array.from({ length: matrix.length }, () => undefined));
  for (let i = 0; i < matrix.length; i++) {
    for (let j = 0; j < matrix[i].length; j++) {
      newList2d[j][matrix.length - 1 - i] = matrix[i][j];
    }
  }
  return newList2d;
}

